### PR TITLE
CI-App: TestNG instrumentation

### PR DESCRIPTION
**Pull request by drodriguezhdez**
**Description**
* Added instrumentation to trace `TestNG v6.4+` test executions.

The instrumentation is applied to the method `initializeDefaultListeners` of the class `org.testng.TestNG` which will be executed during the `TestNG` test framework launching.

Once the method is executed, the `datadog.trace.instrumentation.testng.TracingListener` is registered in that `TestNG` instance.

A `datadog.trace.instrumentation.testng.TestNGDecorator` instance is used in the listener methods to report the traces.

The instrumentation will be disabled by default. This can be enabled configuring the following:

* Environment Variable: `DD_INTEGRATION_TESTNG_ENABLED=true`
* System Property: `-Ddd.integration.testng.enabled=true`

Additionally, some test java classes are created to be used in the tests of this instrumentation. As these classes contain executable tests, it's necessary to exclude them explicitly in the `java.gradle` file to avoid being executed as part of the dd-java-agent build.

### CODE REVIEW

File: dd-java-agent/instrumentation/testng-6.4/src/test/groovy/TestNGTest.groovy [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import org.example.TestSkipped
+import org.example.TestSucceed
+import org.testng.TestNG
+
+class TestNGTest extends TestFrameworkTest {
+
+  def "test success generates spans"() {
+    setup:
+    def testNG = new TestNG()
+    testNG.setTestClasses(TestSucceed)
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/testng-6.4/src/test/groovy/TestNGTest.groovy:17]: Would it be possible to `setOutputDirectory` to `build/tmp/test-output`? Then you wouldn't need to add the line in `.gitignore`.

**Review by bantonsson — APPROVED** [review]

File: dd-java-agent/instrumentation/testng-6.4/src/main/java/datadog/trace/instrumentation/testng/TestNGDecorator.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.testng.ITestResult;
+
+@Slf4j
+public class TestNGDecorator extends TestDecorator {
+  public static final TestNGDecorator DECORATE = new TestNGDecorator();
+
+  @Override
+  public String testFramework() {
+    return "testng";
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/testng-6.4/src/main/java/datadog/trace/instrumentation/testng/TestNGDecorator.java:15]: lol... every time I see this it looks like a typo for `testing`.

File: gradle/java.gradle [hunk 3]
```diff
@@ -278,7 +278,7 @@ def isJdkForced(String javaName) {
 }
 
 def isTestingInstrumentation(Project project) {
-  return project.name == "junit-4.10"
+  return project.name == "junit-4.10" || project.name == "testng-6.4"
```
Comment (tylerbenson) [inline gradle/java.gradle:281]:
```suggestion
  return [ "junit-4.10", "testng-6.4" ].contains(project.name)
```

(or something like this... I assume there will be more testing projects.)

File: settings.gradle [hunk 4]
```diff
@@ -156,6 +156,7 @@ include ':dd-java-agent:instrumentation:spring-scheduling-3.1'
 include ':dd-java-agent:instrumentation:spring-webmvc-3.1'
 include ':dd-java-agent:instrumentation:spring-webflux-5'
 include ':dd-java-agent:instrumentation:spymemcached-2.12'
+include 'dd-java-agent:instrumentation:testng-6.4'
```
Comment (tylerbenson) [inline settings.gradle:159]:
```suggestion
include ':dd-java-agent:instrumentation:testng-6.4'
```

File: dd-java-agent/instrumentation/testng-6.4/src/test/groovy/TestNGTest.groovy (hunk 1 above)
Comment (drodriguezhdez) [reply dd-java-agent/instrumentation/testng-6.4/src/test/groovy/TestNGTest.groovy:19]: Changed 👍

**Pull request merged**